### PR TITLE
Refactoring: Collapse Http1Upstream and Http2Upstream into single HttpUpstream

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
Commit Message: Re-pushing https://github.com/envoyproxy/envoy/pull/30725 without extending upstream lifetime through deferredDelete

This PR is preparation to address https://github.com/envoyproxy/envoy/issues/24724#issuecomment-1377611888.
We are combining Http1Upstream and Http2Upstream into single HttpUpstream because in followup PRs, we would add changes to make this combined HttpUpstream own a Router::UpstreamRequest instance, which would handle both the codecs transparently as in the `Router::Filter` case.
@KBaichoo @alyssawilk Created this new PR as suggested [here](https://github.com/envoyproxy/envoy/pull/27991#issuecomment-1690287653). 

xref:  https://github.com/envoyproxy/envoy/pull/27183

Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by cpakulski — COMMENTED** [review]
Would you mind explaining why is this change required?

I understand that there are some logic differences in handling HTTP1 vs HTTP2+, but it is normal object oriented practice to implement those differences in virtual methods and operate using base class only (basically what is currently implemented).

**Comment by vikaschoudhary16** [comment]
> Would you mind explaining why is this change required?
> 
> I understand that there are some logic differences in handling HTTP1 vs HTTP2+, but it is normal object oriented practice to implement those differences in virtual methods and operate using base class only (basically what is currently implemented).

Reason for combining the logic in single class is that in followup PR handling h1/h2+ will be offloaded to `Router::Upstream` instance which this class would own. Combining the logic in single class will help in cleanly feature flag the future changes. This discussion was done [here](https://github.com/envoyproxy/envoy/pull/27991#issuecomment-1689334521) . In the linked PR, which is closed now, I explored alternative to current i.e not combing existing ones rather created  one more child class, `CombinedUpstream` of `HttpUpstream` and made that class own `Router::UpstreamRequest`. But that approach was turned down by Alyysa and @KBaichoo and was suggested this current path.

File: source/common/tcp_proxy/upstream.h [hunk 1]
```diff
@@ -208,6 +208,7 @@ class HttpUpstream : public GenericUpstream, protected Http::StreamCallbacks {
   };
   DecoderShim response_decoder_;
   Tcp::ConnectionPool::UpstreamCallbacks& upstream_callbacks_;
+  const Http::CodecType type_;
```
Comment (yanavlasov) [inline source/common/tcp_proxy/upstream.h:211]: Is there a specific architectural reason for this change? Previously different behavior was encapsulated in different classes and now it is encapsulated in the same class using state for differentiation.

This change is arguably an antipattern.

**Review by yanavlasov — COMMENTED** [review]
/wait

File: source/common/tcp_proxy/upstream.h (hunk 1 above)
Comment (alyssawilk) [reply source/common/tcp_proxy/upstream.h:211]: the long term plan is to move all of this code to using an upstreamRequest class which as far as I remember was easier with the combined class.   I the prototype is linked somewhere in the issue if you take a look

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**